### PR TITLE
Patrickp microsoft oauth

### PR DESCRIPTION
**Pull request by ppelayo1**
Note to run this you will need to have the client key added to the backend env in graphql_api/lambda/
MICROSOFT_CLIENT_ID=d2dbfc8f-325c-46bf-a3c2-d1f2da795d9f

Updated this pull request
Changing the readme to explain how to set up the backend/frontend for logging in with a Microsoft Account. Also changed the LoginHandler.jsx file to use a Microsoft Client ID from OauthClientID.js in the constant folder.

### :scroll: Description

- Describe the general shape of this PR (new feature? refactor? bug fix? one-line change?)
Authentication thru Microsoft

- Describe what changes are being made
backend/graphql_api/lambda/graphql/resolvers/authentication.js
Removed the issuer as microsoft uses multiple issuers dependent on organization origin. Personal accounts and organizations each have their own unique issuer number.
Corrected signing keys to proper url.
Needed to implement a promise and an await as line 126 was assigning payload properties before the fetch was complete.
Needed to fix the name assignment. Microsoft accounts don't require a name to be assigned to a Microsoft Account and one of my test accounts gave a backend error as a result. The backend will now assign a default first and last name if needed. 
Worth further testing to see if Microsoft allows accounts with only first name or last name, would need to account for those possibilities in the future.

backend/graphql_api/lambda/graphql/resolvers/query.js
Fixing bug. Rest of backend expects a user object.

frontend/src/graphql/graphql.js
Copied the graphql query from the backend for the front end to use

frontend/src/components/LoginHandler/LoginHandler.jsx
Uses React-MIcrosoft-Login to grab an IDToken from Microsoft

frontend/src/utils/verifyToken.js
Implemented a regular expression to do a quick verification of the issuer for Microsoft accounts. Can probably use this later on to verify the backend Microsoft tokens by issuer. 
- Describe why these changes are being made
To let users login thru a Microsoft account of any kind.

- List the use cases and edge cases relevant to this PR
Normal use would just be logging in.

- Describe how errors will be handled. How will we know if this code breaks in production

- Describe any external libraries/dependencies added or removed in this PR
React Microsoft Login was added. This implements a simple call to the microsoft API and returns an IDToken

- Describe any security risks are there regarding this change
Need to verify the issuer in the backend to properly validate the tokens.

- Describe how you tested these changes
I tested with a personal Microsoft account, an account from my school organization, and an account linked to another organization. 

- Link to relevant external documentation
https://github.com/alexandrtovmach/react-microsoft-login

--------------------------
### :clipboard: Mandatory Checklist

- [x] Example of a checked item (please remove when creating your Pull Request) 

- [ ] Linked to the Github Issues being addressed using the right sidebar :arrow_right:
- [ ] Have you discussed these changes with the project leader(s)?
- [ ] Do all variable and function names communicate what they do?
- [ ] Were all the changes commented and / or documented?
- [ ] Is the PR the right size? (If the PR is too large to review, it might be good to break it up into multiple PRs.)
- [ ] Does all work in progress, temporary, or debugger code have a TODO comment with links to Github issues?
- [ ] *If you changed the user interface, did you add before and after screenshots to below?*

--------------------------
#### After

--------------------------
### :blue_book: Glossary
- PR = Pull Request

### CODE REVIEW

File: frontend/src/components/LoginHandler/LoginHandler.jsx [hunk 1]
```diff
@@ -19,6 +20,7 @@ import LoginContext from '../LoginContext/LoginContext';
 // global constant options
 const OPTIONS = {
   googleClientID: '794344810158-sani885h3b9sksk7oqi0cb3spit2271p.apps.googleusercontent.com',
+  microsoftClientID: 'd2dbfc8f-325c-46bf-a3c2-d1f2da795d9f',
```
Comment (JMStudiosJoe) [inline frontend/src/components/LoginHandler/LoginHandler.jsx:23]: these keys should not be here or are they for public use?

**Review by JMStudiosJoe — CHANGES_REQUESTED** [review]
need to know a bout the added key before merging.

File: frontend/src/components/LoginHandler/LoginHandler.jsx (hunk 1 above)
Comment (ppelayo1) [reply frontend/src/components/LoginHandler/LoginHandler.jsx:23]: I need to rewrite code a little to remove it. Going to be done like my other pr up right now by updating Readme.

It is fine for public to know about the key. It is my key so anyone can use it for testing, and a client ID is exposed to the user browser during the redirect anyway. 

It is the secret key we don't want exposed ever. That is only handled by backend.

Comment (ppelayo1) [reply frontend/src/components/LoginHandler/LoginHandler.jsx:23]: So don't approve this pr just yet until I revise it 😀

**Comment by JMStudiosJoe** [comment]
is this pr ready for merging or no?

**Comment by ppelayo1** [comment]
Don't approve this this until I make some revisions; :-).

**Comment by ppelayo1** [comment]
Ready for final review. Updated the readme to explain how to setup for logging in through Microsoft.

**Comment by JMStudiosJoe** [comment]
Oh just saw this will review later today thanks!

**Review by JMStudiosJoe — APPROVED** [review]

**Comment by JMStudiosJoe** [comment]
"later today" -> 16 days later...... sorry.....

**Pull request merged**